libs/dyn/pattern_trie.go

```diff
@@ -0,0 +1,200 @@
+//
+// Note: Setting both key (one of pathKey or anyKey) and index (one of pathIndex or anyIndex)
+// is not supported by the [PatternTrie.SearchPath] method. We don't perform validation for this
+// case because it's not expected to arise in practice where a field is either a map or an array,
+// but not both.
+type trieNode struct {
+	// If set this indicates the trie node is an anyKey node.
+	// Maps to the [AnyKey] component.
+	anyKey *trieNode
+
+	// Indicates the trie node is an anyIndex node.
+	// Maps to the [AnyIndex] component.
+	anyIndex *trieNode
+
+	// Set of strings which this trie node matches.
+	// Maps to the [Key] component.
+	pathKey map[string]*trieNode
+
+	// Set of indices which this trie node matches.
+	// Maps to the [Index] component.
+	pathIndex map[int]*trieNode
+
+	// Indicates if this node is the end of a pattern. Encountering a node
+	// with isEnd set to true in a trie means the pattern from the root to this
+	// node is a complete pattern.
+	isEnd bool
+}
+
+// NewPatternTrie creates a new empty pattern trie.
+func NewPatternTrie() *PatternTrie {
+	return &PatternTrie{
+		root: &trieNode{},
+	}
+}
+
+// Insert adds a pattern to the trie.
+func (t *PatternTrie) Insert(pattern Pattern) error {
+	// Empty pattern represents the root.
+	if len(pattern) == 0 {
+		t.root.isEnd = true
+		return nil
+	}
+
+	current := t.root
+	for i, component := range pattern {
+		// Create next node based on component type
+		var next *trieNode
+		switch c := component.(type) {
+		case anyKeyComponent:
+			if current.anyKey == nil {
+				current.anyKey = &trieNode{}
+			}
+			next = current.anyKey
+
+		case anyIndexComponent:
+			if current.anyIndex == nil {
+				current.anyIndex = &trieNode{}
+			}
+			next = current.anyIndex
+
+		case pathComponent:
+			if key := c.Key(); key != "" {
+				if current.pathKey == nil {
+					current.pathKey = make(map[string]*trieNode)
+				}
+				if _, exists := current.pathKey[key]; !exists {
+					current.pathKey[key] = &trieNode{}
+				}
+				next = current.pathKey[key]
+			} else {
+				idx := c.Index()
+				if current.pathIndex == nil {
+					current.pathIndex = make(map[int]*trieNode)
+				}
+				if _, exists := current.pathIndex[idx]; !exists {
+					current.pathIndex[idx] = &trieNode{}
+				}
+				next = current.pathIndex[idx]
+			}
+		}
+
+		if next == nil {
+			return fmt.Errorf("invalid component type: %T", component)
+		}
+
+		// Mark as end of pattern if this is the last component.
+		if !next.isEnd && i == len(pattern)-1 {
+			next.isEnd = true
+		}
+
+		// Move to next node
+		current = next
+	}
+
+	return nil
+}
+
+// SearchPath checks if the given path matches any pattern in the trie.
+// A path matches if it exactly matches a pattern or if it matches a pattern with wildcards.
+func (t *PatternTrie) SearchPath(path Path) (Pattern, bool) {
+	// We statically allocate the prefix array that is used to track the current
```

Comment:
nit: I'd call this pre-allocate, static allocation means something else https://en.wikipedia.org/wiki/Static_variable